x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushLimitToKnn.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules.logical;
+
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.util.Holder;
+import org.elasticsearch.xpack.esql.expression.function.vector.Knn;
+import org.elasticsearch.xpack.esql.optimizer.LogicalOptimizerContext;
+import org.elasticsearch.xpack.esql.plan.logical.Aggregate;
+import org.elasticsearch.xpack.esql.plan.logical.Filter;
+import org.elasticsearch.xpack.esql.plan.logical.Limit;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.plan.logical.TopN;
+import org.elasticsearch.xpack.esql.plan.logical.inference.Rerank;
+
+/**
+ * Traverses the logical plan and pushes down the limit to the KNN function(s) in filter expressions, so KNN can use
+ * it to set k if not specified.
+ */
+public class PushLimitToKnn extends OptimizerRules.ParameterizedOptimizerRule<Limit, LogicalOptimizerContext> {
+
+    public PushLimitToKnn() {
+        super(OptimizerRules.TransformDirection.DOWN);
+    }
+
+    @Override
+    public LogicalPlan rule(Limit limit, LogicalOptimizerContext ctx) {
+        Holder<Boolean> breakerReached = new Holder<>(false);
+        Holder<Boolean> firstLimit = new Holder<>(false);
+        return limit.transformDown(plan -> {
+            if (breakerReached.get()) {
+                // We reached a breaker and don't want to continue processing
+                return plan;
+            }
+            if (plan instanceof Filter filter) {
+                Expression limitAppliedExpression = limitFilterExpressions(filter.condition(), limit, ctx);
+                if (limitAppliedExpression.equals(filter.condition()) == false) {
+                    return filter.with(limitAppliedExpression);
+                }
+            } else if (plan instanceof Limit) {
+                // Break if it's not the initial limit
+                breakerReached.set(firstLimit.get());
+                firstLimit.set(true);
+            } else if (plan instanceof TopN || plan instanceof Rerank || plan instanceof Aggregate) {
```

Comment:
what happens in the case where we reach TopN/Rerank/Aggregate before we reach a LIMIT plan?
will the value of `k` be null in this case? does that mean we will later fail to translate the knn function in a knn query?

Reply:
@ioanatia That fails as of now - I should have tested for this 😞 . 

I can think of two ways of addressing this:

1) Enforcing using a LIMIT for knn. Fail the query if there's no implicit / explicit LIMIT.

The following would fail:

```
| WHERE knn()
| STATS c = count(*)
```

but also, the following would fail:

```
| WHERE knn()
| STATS c = count(*) where _score > 0.5
```

2) Use an exact search when a LIMIT is not enforced. That would help with the examples above, at the cost of doing a search over all matching rows.

I'm favoring 1) as knn makes no sense if not used as part of a TopN, and users should use an explicit exact search for those cases. It's also a restriction that we can eventually lift if we come with other solution.

WDYT?

Reply:
works for now - we can always address this in another way later 👍